cupy/core/raw.pyx

```diff
@@ -72,41 +74,29 @@ def _get_raw_kernel(code, name, options=()):
     return module.get_function(name)
 
 
-@cupy.util.memoize(for_each_device=True)
-def _get_func_attributes(func):
-    cdef:
-        int sharedSizeBytes, constSizeBytes, localSizeBytes
-        int maxThreadsPerBlock, numRegs, ptxVersion, binaryVersion
-        int cacheModeCA, maxDynamicSharedSizeBytes, preferredShmemCarveout
-
-    sharedSizeBytes = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_SHARED_SIZE_BYTES, func)
-    constSizeBytes = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_CONST_SIZE_BYTES, func)
-    localSizeBytes = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_LOCAL_SIZE_BYTES, func)
-    maxThreadsPerBlock = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_MAX_THREADS_PER_BLOCK, func)
-    numRegs = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_NUM_REGS, func)
-    ptxVersion = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_PTX_VERSION, func)
-    binaryVersion = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_BINARY_VERSION, func)
-    cacheModeCA = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_CACHE_MODE_CA, func)
-    maxDynamicSharedSizeBytes = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES, func)
-    preferredShmemCarveout = driver.funcGetAttribute(
-        driver.CU_FUNC_ATTRIBUTE_PREFERRED_SHARED_MEMORY_CARVEOUT, func)
-
-    return dict(sharedSizeBytes=sharedSizeBytes,
-                constSizeBytes=constSizeBytes,
-                localSizeBytes=localSizeBytes,
-                maxThreadsPerBlock=maxThreadsPerBlock,
-                numRegs=numRegs,
-                ptxVersion=ptxVersion,
-                binaryVersion=binaryVersion,
-                cacheModeCA=cacheModeCA,
-                maxDynamicSharedSizeBytes=maxDynamicSharedSizeBytes,
-                preferredShmemCarveout=preferredShmemCarveout)
+class FunctionAttributes:
```

Comment:
I think this is a better class implementation than the minimal one I had previously proposed on the Cython side. I think the only main advantage of the dictionary approach is that it is a convenient way to quickly see the values for all attributes, but probably reading/setting a particular attribute is a more common scenario.

I guess an alternative to this solution is to leave read attributes in a dictionary, but add setter methods for the two writeable attributes to the RawKernel class itself. I think this is a cleaner and pythonic approach, though.

Also, once attributes can be set the `memoize` approach I had used on the attributes dictionary becomes problematic as the memo would then be out of date once an attribute is changed.

Reply:
OK, I will just leave everything as is pending additional feedback. Thanks again for #1874 you already figured out the hard stuff and that saved me a lot of trouble.